File: python/interviewbit/backtracking/reverse_link_list_recursion/reverse_link_list_recursion.py
'''
https://www.interviewbit.com/problems/reverse-link-list-recursion/

Reverse a linked list using recursion.

Example :
Given 1->2->3->4->5->NULL,

return 5->4->3->2->1->NULL.
'''



'''
Solution Outline:
	Use the call stack to store the SLL nodes in reverse order
	reverse the links on each stack unwind
'''



# Definition for singly-linked list.
class ListNode:
	def __init__(self, x):
		self.val = x
		self.next = None

	# Build a linked-list from a list
	@staticmethod
	def fromList(l):
		head = ListNode(l[0])
		tmp = head
		for x in l[1:]:
			node = ListNode(x)
			tmp.next = node
			tmp = tmp.next
		return head


	# Build a list from a linked-list
	def toList(self):
		tmp = self
		l = []
		while tmp:
			l.append(tmp.val)
			tmp = tmp.next

		return l

		
	def __str__(self):
		return str(self.val)



class Solution:
	# @param A : head node of linked list
	# @return the head node in the linked list
	def reverseList(self, head):
		def reverse_(node):
			if not node.next:
				new_head[0] = node
				return node

			reverse_(node.next).next = node
			return node

		if not head:
			return head

		new_head = [None]
		reverse_(head).next = None
		return new_head[0]


if __name__ == '__main__':
	s = Solution()
	ll = ListNode.fromList([1,2,3,4,5])
	ll = s.reverseList(ll)
	assert ll.toList() == [5,4,3,2,1]

File: c++/interviewbit/backtracking/reverse_link_list_recursion/reverse_link_list_recursion.cc
/**
https://www.interviewbit.com/problems/reverse-link-list-recursion/

Reverse a linked list using recursion.

Example :
Given 1->2->3->4->5->NULL,

return 5->4->3->2->1->NULL.
*/



/**
Solution Outline:
	Use the call stack to store the SLL nodes in reverse order
	reverse the links on each stack unwind
*/


#include <vector>
#include <iostream>
#include <cassert>

/**
 * Definition for singly-linked list.
 */
struct ListNode {
	int val;
	ListNode *next;
	ListNode(int x) : val(x), next(NULL) {}

	/** Convert a linked list into a vector of integers */
	std::vector<int> toList() {
		std::vector<int> v;
		ListNode *ll = this;
		while (ll) {
			v.push_back(ll->val);
			ll = ll->next;
		}
		return v;
	}


	/** Create a linked list from a vector of integers */
	static ListNode *fromList(std::vector<int> v) {
		ListNode *head = NULL;
		ListNode *tail = NULL;
		for (auto x: v) {
			ListNode *node = new ListNode(x);
			if (!head) {
				head = tail = node;
			} else {
				tail->next = node;
				tail = node;
			}
		}
		return head;
	}

};

class Solution {
	public:
		ListNode *reverseList(ListNode *head);
		ListNode *reverse_(ListNode *node, ListNode **new_head) {
			if (!node->next) {
				/** Found the tail node, make it the new_head */
				*new_head = node;
				return node;
			}

			reverse_(node->next, new_head)->next = node;
			return node;
		}
};


/** Overload operator << so cout<<linked list works */
std::ostream& operator<<(std::ostream &os, ListNode *ll) {
	while (ll) {
		os << ll->val << " ";
		ll = ll->next;
	}

	return os;
}

/** reverse a linked list, head, in-place and return the new head */
ListNode*
Solution::reverseList(ListNode* head) {
	if (!head)
		return 0;

	ListNode *new_head;
	this->reverse_(head, &new_head)->next = 0;
	return new_head;
}


int main(void)
{
	Solution s;
	std::vector<int> v = {1,2,3,4,5};
	ListNode *ll = ListNode::fromList(std::vector<int>({1,2,3,4,5}));
	assert(ll->toList() == std::vector<int>({1,2,3,4,5}));
	ll = s.reverseList(ll);
	assert(ll->toList() == std::vector<int>({5,4,3,2,1}));
}

